File: src/2025-03-04-03-28-40--deepseek-chat-0304/python/exercises/practice/poker/poker.py
def best_hands(hands):
    """Pick the best hand(s) from a list of poker hands.
    
    Args:
        hands: List of strings representing poker hands
        
    Returns:
        List of strings containing the best hand(s)
    """
    if not hands:
        return []
        
    # Parse all hands and calculate their rank values
    parsed_hands = []
    for hand in hands:
        cards = hand.split()
        values = sorted((card_value(card[0]) for card in cards), reverse=True)
        suits = [card[-1] for card in cards]
        is_flush = len(set(suits)) == 1
        is_straight = is_consecutive(values)
        
        # Count value occurrences for pairs, three/four of a kind
        value_counts = {}
        for v in values:
            value_counts[v] = value_counts.get(v, 0) + 1
        sorted_counts = sorted(value_counts.items(), key=lambda x: (x[1], x[0]), reverse=True)
        
        # Determine hand rank
        if is_straight and is_flush:
            # Special case for low straight flush (A-2-3-4-5)
            if values == [14, 5, 4, 3, 2]:
                rank = (8, [5])  # Treat as 5-high straight flush
            else:
                rank = (8, values)  # Straight flush
        elif sorted_counts[0][1] == 4:
            rank = (7, [sorted_counts[0][0]] + [x[0] for x in sorted_counts[1:]])  # Four of a kind
        elif sorted_counts[0][1] == 3 and sorted_counts[1][1] == 2:
            rank = (6, [sorted_counts[0][0], sorted_counts[1][0]])  # Full house
        elif is_flush:
            rank = (5, values)  # Flush
        elif is_straight:
            rank = (4, values)  # Straight
        elif sorted_counts[0][1] == 3:
            rank = (3, [sorted_counts[0][0]] + [x[0] for x in sorted_counts[1:]])  # Three of a kind
        elif sorted_counts[0][1] == 2 and sorted_counts[1][1] == 2:
            rank = (2, sorted([sorted_counts[0][0], sorted_counts[1][0]], reverse=True) + 
                    [x[0] for x in sorted_counts[2:]])  # Two pair
        elif sorted_counts[0][1] == 2:
            rank = (1, [sorted_counts[0][0]] + [x[0] for x in sorted_counts[1:]])  # One pair
        else:
            rank = (0, values)  # High card
            
        parsed_hands.append((rank, hand))
    
    # Find the maximum rank
    max_rank = max(parsed_hands, key=lambda x: x[0])[0]
    
    # Return all hands with the maximum rank
    return [hand for (rank, hand) in parsed_hands if rank == max_rank]

def card_value(card):
    """Convert card rank to numerical value."""
    if card == 'J':
        return 11
    elif card == 'Q':
        return 12
    elif card == 'K':
        return 13
    elif card == 'A':
        return 14
    elif card == 'T':
        return 10
    else:
        return int(card)

def is_consecutive(values):
    """Check if values are consecutive, handling Aces as both high and low."""
    # Handle low straight (A-2-3-4-5)
    if values == [14, 5, 4, 3, 2]:
        return True
    # Handle normal consecutive
    return all(values[i] == values[i+1] + 1 for i in range(len(values)-1))

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/poker/poker.js
// Helper functions
const getRankValue = (rank) => {
  const ranks = '23456789TJQKA';
  return ranks.indexOf(rank);
};

const getHandDetails = (hand) => {
  const cards = hand.split(' ');
  const faces = cards.map(card => card[0]);
  const suits = cards.map(card => card[1]);
  
  // Count duplicates and get face values
  const counts = faces.reduce((acc, face) => {
    acc[face] = (acc[face] || 0) + 1;
    return acc;
  }, {});
  
  // Sort faces by count first, then by rank
  const sortedFaces = Object.keys(counts).sort((a, b) => {
    const countDiff = counts[b] - counts[a];
    if (countDiff !== 0) return countDiff;
    return getRankValue(b) - getRankValue(a);
  });
  
  // Get all face values sorted by rank for tie-breakers
  const faceValues = faces.map(face => getRankValue(face)).sort((a, b) => b - a);
  
  // Check for flush
  const flush = suits.every(suit => suit === suits[0]);
  
  // Check for straight
  const uniqueSorted = [...new Set(faceValues)].sort((a, b) => b - a);
  const isStraight = uniqueSorted.length === 5 && 
    (uniqueSorted[0] - uniqueSorted[4] === 4 || 
     (uniqueSorted.includes(12) && uniqueSorted.includes(3))); // Handle A-2-3-4-5
  
  return {
    counts,
    sortedFaces,
    faceValues,
    flush,
    straight: isStraight,
    isLowStraight: isStraight && uniqueSorted.includes(3) // A-2-3-4-5
  };
};

const getHandStrength = (handDetails) => {
  const { counts, sortedFaces, faceValues, flush, straight, isLowStraight } = handDetails;
  
  // Get counts in order
  const countValues = Object.values(counts).sort((a, b) => b - a);
  
  // Adjust for low straight
  const adjustedValues = isLowStraight ? [3, 2, 1, 0, 12] : faceValues;
  
  if (flush && straight) {
    return { strength: 8, values: adjustedValues }; // Straight flush
  }
  if (countValues[0] === 4) {
    return { strength: 7, values: [getRankValue(sortedFaces[0]), ...adjustedValues] }; // Four of a kind
  }
  if (countValues[0] === 3 && countValues[1] === 2) {
    return { strength: 6, values: [getRankValue(sortedFaces[0]), getRankValue(sortedFaces[1])] }; // Full house
  }
  if (flush) {
    return { strength: 5, values: adjustedValues }; // Flush
  }
  if (straight) {
    return { strength: 4, values: adjustedValues }; // Straight
  }
  if (countValues[0] === 3) {
    return { strength: 3, values: [getRankValue(sortedFaces[0]), ...adjustedValues] }; // Three of a kind
  }
  if (countValues[0] === 2 && countValues[1] === 2) {
    return { 
      strength: 2, 
      values: [
        getRankValue(sortedFaces[0]), 
        getRankValue(sortedFaces[1]),
        ...adjustedValues.filter(v => v !== getRankValue(sortedFaces[0]) && v !== getRankValue(sortedFaces[1]))
      ]
    }; // Two pair
  }
  if (countValues[0] === 2) {
    return { 
      strength: 1, 
      values: [
        getRankValue(sortedFaces[0]),
        ...adjustedValues.filter(v => v !== getRankValue(sortedFaces[0]))
      ]
    }; // One pair
  }
  return { strength: 0, values: adjustedValues }; // High card
};

const compareHands = (hand1, hand2) => {
  const details1 = getHandDetails(hand1);
  const details2 = getHandDetails(hand2);
  const strength1 = getHandStrength(details1);
  const strength2 = getHandStrength(details2);
  
  // Compare hand strengths
  if (strength1.strength !== strength2.strength) {
    return strength2.strength - strength1.strength;
  }
  
  // Compare values for tie-breakers
  for (let i = 0; i < strength1.values.length; i++) {
    if (strength1.values[i] !== strength2.values[i]) {
      return strength2.values[i] - strength1.values[i];
    }
  }
  
  return 0;
};

export const bestHands = (hands) => {
  if (hands.length === 1) return hands;
  
  let best = [hands[0]];
  for (let i = 1; i < hands.length; i++) {
    const comparison = compareHands(best[0], hands[i]);
    if (comparison < 0) {
      best = [hands[i]];
    } else if (comparison === 0) {
      best.push(hands[i]);
    }
  }
  
  return best;
};
